### PR TITLE
gh-120176: Reduce number of compares in index range checks

### LINKED ISSUE #120176: Use shared helper function for faster index range checks
#9784 (discussed in #72583 and #67741) introduced a helper function to reduce the number of comparisons in list index range checks (see https://godbolt.org/z/jM8oexT1d).

The same pattern can be applied in many other places in the code base like in `arraymodule.c`, `mmapmodule.c`, `bytearrayobject.c`, etc.

I'd propose to extract this helper function into a header and start using it in more places for index range checks. [`internal/pycore_abstract.h`](https://github.com/python/cpython/blob/main/Include/internal/pycore_abstract.h) seems like it might be a decent place to put this helper in.

I'm happy to make a PR if people think a micro optimisation like this justifies the code churn, but let me know if I'm missing something.

### Linked PRs
* gh-120181

### PR DESCRIPTION
**Pull request by lgeiger**
This PR is a followup to #9784 (discussed in #72583 and #67741) and extracts the helper function into `internal/pycore_abstract.h` and uses it where applicable to reduce the number of comparisons in index range checks.

See https://godbolt.org/z/jM8oexT1d for an assembly comparison.

I feel like this change doesn't reduce readability but let me know if such micro optimisation does not justify the code churn.

* Issue: gh-120176

### CODE REVIEW

**Comment by methane** [comment]
No news entry is required because this change is not visible to users.
I assume performance gain is not large enough to have news entry.

**Review by methane — APPROVED** [review]
LGTM.
I will wait a week before merging this PR to ensure that other committers have the opportunity to review it.

**Comment by eendebakpt** [comment]
@lgeiger @methane This pr is nearly identical to https://github.com/python/cpython/pull/99762. There are some more review comments there

**Comment by serhiy-storchaka** [comment]
This idea was already discussed and rejected. See #72583.

**Comment by lgeiger** [comment]
Ah sorry, didn't see #99762 

Sorry about that!

**Pull request closed without merging**
